### ISSUE TITLE
Add native support for Pydantic parameters to Kedro

### ISSUE THREAD

**Issue by datajoely** [kedro-org/kedro#5110]
## Introduction

Introduce opt-in, backwards compatible support for Pydantic models to define, validate, and expose Kedro parameters. This improves type safety, developer experience, and would unlock integration with APIs, UI tooling, and external orchestration systems. Hide new and exciting features behind this pattern to boost adoption.

## Background

Currently, Kedro parameters are YAML dicts with no validation or typing. This leads to brittle pipelines, runtime errors, and poor ergonomics. Many ecosystems already use Pydantic for structured configuration:

* **Prefect**: [Data validation with Pydantic](https://www.prefect.io/blog/data-validation-with-pydantic)
* **OpenAI**: [Structured outputs in the API](https://openai.com/index/introducing-structured-outputs-in-the-api/)
* **Ollama**: [Structured outputs](https://ollama.com/blog/structured-outputs)

Pydantic generates JSON Schema automatically, making it trivial to generate expert user interfaces from this pattern:

* **JSON Schema → UI**: [react-jsonschema-form](https://rjsf-team.github.io/react-jsonschema-form/)

Imagine a world where you you got a rich slider component from defining a config parameter like this:

```python
class Example(BaseModel):
    value: int = Field(ge=0, le=10)
```

## Problem

* **Implicit contracts**: Node functions depend on parameter shapes that are only informally documented.
* **Brittle configs**: Typos or missing fields fail only at runtime.
* **Weak tooling**: No autocompletion, type hints, or schema-based validation.
* **Integration gaps**: APIs and MCPs require JSON Schema for strong contracts; Kedro provides none.

### What's in scope

* Allow projects to declare Pydantic models for parameter namespaces.
* Validate YAML parameters against models at context creation. Fail fast with invalid CLI arguemnts.
* Inject typed model instances into nodes.
* Export JSON Schema (`kedro params schema`).
* Provide scaffolding CLI (`kedro params scaffold`).

### What's not in scope

* Replacing YAML entirely.
* Removing the traditional (unsafe) pattern.

### 1. YAML parameters (`conf/base/parameters/training.yml`)

```yaml
learning_rate: 0.05
epochs: 5
```

---

### 2. Node function with a typed parameter

```python
from pydantic import BaseModel, Field

class TrainingParams(BaseModel):
    learning_rate: float = Field(ge=0, le=1)
    epochs: int = Field(ge=1, le=10)

def train_model(X, params: TrainingParams):
    # params is validated automatically before reaching here
    print(f"Training for {params.epochs} epochs at lr={params.learning_rate}")
    return "trained_model"
```

---

### 3. Pipeline definition

```python
from kedro.pipeline import node, Pipeline

def create_pipeline() -> Pipeline:
    return Pipeline(
        [
            node(
                func=train_model,
                inputs=["X", "params:training"],  # standard Kedro param binding
                outputs="trained_model",
            )
        ]
    )
```

---

**What happens**

* Kedro loads `conf/base/parameters/training.yml` as a dict.
* The pipeline object sees `params: TrainingParams` annotation on `train_model`.
* It converts the dict → `TrainingParams` (validating along the way).
* Inside `train_model`, you work with a fully typed object, not a raw dict.

👉 From the user’s perspective: just **define a model, annotate the function, and use YAML as usual**. Everything else is handled automatically.

## Alternatives

* **Use stdlib** `dataclass`: no native JSON Schema or descriptions.
* **Do nothing**: leaves parameters untyped, brittle, and harder to integrate.

## Testing

* Unit: schema validation success/failure, error messages.
* Integration: typed injection, CLI schema export, scaffold.
* Contract: JSON Schema consumption by RJSF.
* Performance: compare context creation times.

## Rollout strategy

* Backward compatible: plugin or core functionality is opt-in.
* Update docs to show this as preferred approach
* Coinside with Viz features that are only possible with this pattern.

## Future iterations

* Expose parameter schemas in APIs/MCPs.
* Example UI with react-jsonschema-form.
* Store validated parameter provenance.
* Extend schema support to catalog/hooks if demand exists.

Related Issues - https://github.com/kedro-org/kedro/issues/328

**Comment by noklam** [comment]
One immediate thing that may become a problem is that:
```python
def train_model(X, params: TrainingParams):
    # params is validated automatically before reaching here
    print(f"Training for {params.epochs} epochs at lr={params.learning_rate}")
    return "trained_model"
```

This now force the type in the function, I don't particular like this because:
1. It's feels a bit weird to force a BaseModel class as function args - you immediately loss the Python signature and kinda hide it in a slightly better documented **kwargs.
2. The function is now bounded to a class - and you cannot reuse the function elsewhere.

**Comment by deepyaman** [comment]
It would be nice to not force a dependency on Pydantic just for this. I think if it's an extra/works if Pydantic is installed, that's fine. You could also consider supporting dataclass annotations later then, if so desired.

> Update docs to show this as preferred approach

I think this should depend on demand, or there should be at least a very clear set of additional functionality (e.g. whatever you mention in Kedro-Viz) that would justify it. Definitely feels like a more "advanced" feature if you have to learn about Pydantic models to get started with Kedro.

> One immediate thing that may become a problem is that:
> 
> def train_model(X, params: TrainingParams):
>     # params is validated automatically before reaching here
>     print(f"Training for {params.epochs} epochs at lr={params.learning_rate}")
>     return "trained_model"
> This now force the type in the function, I don't particular like this because:
> 
> 1. It's feels a bit weird to force a BaseModel class as function args - you immediately loss the Python signature and kinda hide it in a slightly better documented **kwargs.
> 2. The function is now bounded to a class - and you cannot reuse the function elsewhere.

@noklam Is this a problem given it's opt-in? It seems like somebody who wants to use Pydantic optionally could.

Last but not least, may be out of scope of this proposal, but this feels like it could be a part of broader validation initiatives that extend beyond parameters. For example, why not enable one (or two) steps further?

At the very least, it should be possible to also check the input type to a node:

```python
import pandas as pd
from pydantic import BaseModel, Field


class TrainingParams(BaseModel):
    learning_rate: float = Field(ge=0, le=1)
    epochs: int = Field(ge=1, le=10)


def train_model(X: pd.DataFrame, params: TrainingParams):
    # X's data type is validated automatically before reaching here
    # params is validated automatically before reaching here
    print(f"Training for {params.epochs} epochs at lr={params.learning_rate}")
    return "trained_model"
```

But why not also enable valdiation in the same way?

```python
import pandas as pd
import pandera.pandas as pa
from pandera.typing.pandas import Index, DataFrame, Series
from pydantic import BaseModel, Field


class InputSchema(pa.DataFrameModel):
    year: Series[int] = pa.Field(gt=2000, coerce=True)
    month: Series[int] = pa.Field(ge=1, le=12, coerce=True)
    day: Series[int] = pa.Field(ge=0, le=365, coerce=True)


def train_model(X: pd.DataFrame[InputSchema], params: TrainingParams):
    # X's content is validated automatically before reaching ehre
    # params is validated automatically before reaching here
    print(f"Training for {params.epochs} epochs at lr={params.learning_rate}")
    return "trained_model"
```

**Comment by datajoely** [comment]
@noklam the function signature thing is interesting. I don't know if I agree. 

> It's feels a bit weird to force a BaseModel class as function args - you immediately loss the Python signature and kinda hide it in a slightly better documented **kwargs.

<img width="282" height="199" alt="Image" src="https://github.com/user-attachments/assets/9851471c-f239-439d-ac1b-5f087cd9f562" />

Even without LLM autocomplete, function signatures are helpful here (example using Cursor/VS Code).

> The function is now bounded to a class - and you cannot reuse the function elsewhere.

The coupling of the python function to a model. This is interesting, my view your python functions should exist in an already separate, independently tested package. In this situation you should be able to import this gracefully, if we're talking about `nodes.py` you're right it becomes a bit more hairy. 

What do you think?

----

@deepyaman , thanks for your thoughts here

> It would be nice to not force a dependency on Pydantic just for this. I think if it's an extra/works if Pydantic is installed, that's fine. You could also consider supporting dataclass annotations later then, if so desired.

I'm in two minds here. dataclasses/typedicts give you the types but they don't give you:

- Validators - at least without more work
- Field descriptions - which will be useful potentially for any auto-generated documentation
- JSON serialisation / JSONSchema validation

I should also say Pydantic is a core part of Kedro-Viz so at the very least we expect people to have this dependency and it to compile with the rest of your project. If we want to throw it in an optional group fine, but I still think it should be part of the 'golden path' we introduce right from the tutorial 

> Last but not least, may be out of scope of this proposal, but this feels like it could be a part of broader validation initiatives that extend beyond parameters. For example, why not enable one (or two) steps further?

Now this is where I would encourage pure, native Pandera? Especially now it supports all of our preferred dataframe types.

**Comment by astrojuanlu** [comment]
Past discussion https://github.com/kedro-org/kedro/discussions/3788

**Comment by astrojuanlu** [comment]
In my `~/Projects/Skunkworks` folder I have this, I can release it if there's interest.

---

## Quickstart

Define your YAML file as follows:

```yaml
# catalog.yml
input_1:
  type: datasets.EagerPolarsDataset
  path: s3://bucket/my_dataset.parquet

sensors:
  type: parameters.Sensors
  sensor1:
    name: "Sensor 1"
  sensor2:
    name: "Sensor 2"
    type: "temperature"
    stderr: 0.1
```

Define your parameter models:

```python
# parameters.py
import typing as t

from pydantic import BaseModel, Field, NonNegativeFloat, RootModel

class Sensor(BaseModel):
    name: str
    type_: t.Annotated[str, Field(alias="type")] = "unknown"
    stderr: t.Annotated[float, NonNegativeFloat] = 0.0

class Sensors(RootModel):
    root: dict[str, Sensor]
```

Define your datasets:

```python
# datasets.py
import typing as t
from dataclasses import dataclass

from pydantic import BaseModel
from pydantic.networks import AnyUrl
import polars as pl

class EagerPolarsDataset(BaseModel):
    path: AnyUrl


@dataclass
class EagerPolarsDatasetIO:
    spec: EagerPolarsDataset

    @property
    def path(self) -> str:
        return str(self.spec.path)

    def load(self) -> pl.DataFrame:
        return pl.read_parquet(self.path)

    def save(self, data: pl.DataFrame) -> None:
        data.write_parquet(self.path)
```

And finally, instantiate your catalog:

```
>>> from kedrong4_catalog import DataCatalog
>>> from parameters import Sensors
>>> from datasets import EagerPolarsDatasetIO
>>> from yaml import safe_load
>>> with open("catalog.yml") as fh:
...     catalog_conf = safe_load(fh)
...     
>>> catalog = DataCatalog.from_config(
...     catalog_conf,
...     specs={
...         "datasets.EagerPolarsDataset": EagerPolarsDataset,
...         "parameters.Sensors": Sensors,
...     },
...     datasets_mapping={
...         EagerPolarsDataset: EagerPolarsDatasetIO,
...     },
... )
>>> catalog.list()
['input_1', 'sensors']
>>> type(catalog.load("input_1"))
polars.DataFrame
>>> catalog.load("sensors")
{'sensor1': {'name': 'Sensor 1', 'type': 'unknown', 'stderr': 0.0}, 'sensor2': {'name': 'Sensor 2', 'type': 'temperature', 'stderr': 0.1}}
```

**Comment by deepyaman** [comment]
> If we want to throw it in an optional group fine, but I still think it should be part of the 'golden path' we introduce right from the tutorial

Seems fine. I think starting as an extra is great, just because validating type hints is a bit more advanced (e.g. looking at the users who are completely new to SWE), but no issue with building up to this in the recommended path.

**Comment by chrisschopp** [comment]
I like the use of a typed object in place of a dict. When a parameter has just one field though, I'm not yet sure how I'd prefer to name it.

Using a node from a recent Kedro project, here is what this could look like:

```python
# parameters.py
import fitter
from pydantic import BaseModel, Field, field_validator


class MinObservations(BaseModel):
    n: int = Field(
        ge=2,
        description=(
            "Minimum number of observations required to try fitting distributions. "
            "Must be at least 2 or `fitter` will throw an error."
        ),
    )


class Distributions(BaseModel):
    to_fit: list[str] = Field(description="Distributions to try fitting to daily demand.")

    @field_validator("to_fit")
    @classmethod
    def validate_distributions(cls, distributions):
        """Fitter is limited to scipy's continuous distributions."""
        invalid_distributions = set(distributions).difference(
            set(fitter.get_distributions())
        )
        if invalid_distributions:
            raise ValueError(
                f"Invalid distributions found: {invalid_distributions}. "
                f"Must be from: {fitter.get_distributions()}"
            )
        return distributions
```

```python
# nodes.py
import polars as pl
from polars import LazyFrame
from .parameters import Distributions, MinObservations


def fit_demand_distributions(
    daily_demand: LazyFrame,
    min_observations: MinObservations,
    distributions: Distributions,
) -> LazyFrame:
    demand = daily_demand.filter(
        pl.col("quantity").len() >= min_observations.n,
        pl.col("quantity").n_unique() >= min_observations.n,
    )
    fit_some_distributions(demand, distributions.to_fit)
```

Maybe it would be better to make the field name redundant with the class name? E.g., the node would contain `min_observations.min_observations` and `distributions.distributions`.

**Comment by datajoely** [comment]
This is great to see @chrisschopp - it would be great to understand if you'd appreciate if Kedro could interpret this parameter argument and:

- Give you pre-flight validation that your YAML or CLI args are valid?
- Generate a UI so you could kick off a run with a slider giving you 1-2 distructions?

**Comment by chrisschopp** [comment]
> * Give you pre-flight validation that your YAML or CLI args are valid?

Absolutely. Better to fail fast instead of when the node runs with an invalid argument.

> * Generate a UI so you could kick off a run with a slider giving you 1-2 distructions?

Very useful for non-technical users who need to run the pipeline. Would I use it for a pipeline I built? I _think_ so. I wouldn't have to remember the parameter names and their allowable values. Would also be helpful with onboarding developers to the project.

---
For some more context:

For a recent Kedro project I implemented parameter validation because the end-users are non-technical. They wanted a UI but we didn't have time/budget to deliver a UI in that phase. I wanted them to have a friendly error message if they had a typo when editing `parameters.yml`. We also wrapped the three distinct pipelines in [Winpython](https://github.com/winpython/winpython) executables so they could run a pipeline by double-clicking a `.bat` file.

`Field` metadata would be helpful for documenting what a parameter is for. If the `description` was displayed in the UI there would be less need for standalone documentation. And fewer things to update when changes occur.

**Comment by rashidakanchwala** [comment]
I think validation is necessary; it’s a software engineering best practice and something Kedro holds itself to. I do like the Pydantic-based validation and it’s a very good point that it’s already a core dependency for Kedro-Viz. I agree with @deepyaman - we should make Pydantic optional; when Kedro-Viz upgraded to Pydantic v2, we did have a handful of users who were still on v1 and couldn’t use viz anymore.

**Comment by datajoely** [comment]
We can talk about this tomorrow at TSC of implementation detail, but I think it would be perfectly fine to recommend it as golden path and build functionality that is dependant on Pydantic being available.

**Comment by DimedS** [comment]
Many thanks @datajoely for today’s tech design discussion! 🙌  
You can find a brief summary of the proposal below.

🎥 **Recording:**  
[Zoom video link](https://zoom.us/rec/play/_KqGArJXcBBWPG8ltI-Lgnclz3Dzz_YQk_LqcntmXXQQQ5P3PpEH7ZfSCRl3MqtAdEulcyVT66g8Jwpi.czS6uZ47laUU5JKW?eagerLoadZvaPages=sidemenu.billing.plan_management&accessLevel=meeting&canPlayFromShare=true&from=share_recording_detail&continueMode=true&componentName=rec-play&originRequestUrl=https%3A%2F%2Fzoom.us%2Frec%2Fshare%2Fk65PZQB68pEutehg9mrJPitlhC8Q3QeJy8brK5enCcMaTDt825sB3fqWvD8-WSoP.h6dsNrydRDQ-1saQ)

🧭 **Next steps:**  
We’ll proceed with a [technical spike (#5161)](https://github.com/kedro-org/kedro/issues/5161) to explore implementation options and performance implications.  
Once the spike is complete, we’ll return with a formal **TSC voting proposal**.

---

### 🧩 Final Proposal Summary — “Add Native Support for Pydantic Parameters”

**Goal:**  
Introduce **optional, native Pydantic-based parameter validation** in Kedro to make configuration safer, more type-safe, and more compatible with APIs, UIs, and LLM/MCP use cases.

**Key points:**
- **Opt-in & backwards compatible.** YAML remains the source of truth; users can continue using plain dicts.  
- **Pydantic models** can be declared and bound to node functions to:
  - Validate YAML parameters early (on `KedroSession` creation).  
  - Fail fast on invalid configs instead of runtime errors.  
  - Provide **rich metadata** (types, descriptions, ranges, validators).  
  - Auto-generate **JSON Schemas**, enabling UIs (e.g., Kedro-Viz sliders/forms) and LLM integrations.  
- **Golden path:** Recommended for new projects and tutorials, but kept **optional** to avoid breaking changes.  
- **Long-term benefits:** Enables structured Kedro-as-a-Service APIs, MCP compatibility, and better UX for non-technical users.  

**Not in scope:**  
Replacing YAML entirely or removing the existing pattern — YAML stays, but gets validated.
